### PR TITLE
Support to hdf on Dxp and added QE65000

### PR DESCRIPTION
**Pull request by gabrielfedel**
Hello, 
I added support to save hdf on mapping process for DxpClass.
I added too a class to spectrometer QE65000 (with hdf support too).

### CODE REVIEW

File: py4syn/epics/DxpClass.py [hunk 1]
```diff
@@ -17,14 +17,17 @@
...
+import h5py
 
 
 class Dxp(StandardDevice, ICountable):
     # CONSTRUCTOR OF DXP CLASS
-    def __init__(self, mnemonic, output, numberOfChannels=4, numberOfRois=32,
-                 pv=None, dxpType="mca", responseTimeout=15):
+    def __init__(self, mnemonic, numberOfChannels=4, numberOfRois=32,
+                 pv=None, dxpType="mca", responseTimeout=15, output="./out",
+                 imageDeep=2048):
```
Comment (hhslepicka) [inline py4syn/epics/DxpClass.py]: What is this ```imageDeep```?
Is this defined once in the device and remain the same for the whole use of it?
The name is a bit confusing.

File: py4syn/epics/DxpClass.py [hunk 2]
```diff
@@ -17,14 +17,17 @@
...
         """ Constructor
         responseTimeout : how much time to wait dxp answer
+        imageDeep : how many points are collected each time
```
Comment (hhslepicka) [inline py4syn/epics/DxpClass.py]: I would recommend to name it something more clear such as ```num_points```.

File: py4syn/epics/DxpClass.py [hunk 3]
```diff
@@ -104,25 +113,42 @@ def getValueChannel(self, **kwargs):
         channel = kwargs['channel']
         c = int(channel[3]) - 1
         if(len(channel) > 4):
-            r = int(channel[5])
-            return self.pvDxpRois[c][r]
+            r = int(channel[6])
```
Comment (hhslepicka) [inline py4syn/epics/DxpClass.py]: How is this ```getValueChannel``` being used?
Why do we need to have this index 6 hardcoded? 
If it is necessary, I would recommend the use of a constant in order to make it easier to understand now and in the future.

File: py4syn/epics/DxpClass.py [hunk 4]
```diff
@@ -104,25 +113,42 @@ def getValueChannel(self, **kwargs):
...
-        if(fileName):
-            spectrum = self.pvDxpChannels[ch].get(as_numpy=True)
-            prefix = fileName.split('.')[0]
-            while os.path.exists('%s_%s%d_%04d.mca' % (prefix, self.dxpType,
-                                                       ch, idx)):
-                idx += 1
-            fileName = '%s_%s%d_%04d.mca' % \
-                       (prefix, self.dxpType, ch, idx)
-        np.savetxt(fileName, spectrum, fmt='%f')
+        self.ch = ch
```
Comment (hhslepicka) [inline py4syn/epics/DxpClass.py]: When is saveSpectrum being called?

Comment (hhslepicka) [reply py4syn/epics/DxpClass.py]: This method is a bit confusing, would you mind to show a code snippet with a use case for it?

File: py4syn/epics/DxpClass.py [hunk 5]
```diff
@@ -190,3 +215,70 @@ def close(self):
...
+
+        # last collected point
+        self.lastPos = 0
+
+    def stopCollectImage(self):
+        """Stop collect image"""
+        self.fileResult.close()
+        self.lastPos = -1
+
+    def setNormValue(self, value):
```
Comment (hhslepicka) [inline py4syn/epics/DxpClass.py]: The filename + saving to txt code appears in more than one place.
Would you mind to refactor it and make it usable for all the cases without repeating the code?

File: py4syn/epics/QE65000Class.py [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+
+    def startCount(self):
+        """ Starts acquiring an spectrum
+        It's necessary to call setCounTime before"""
+
+        if self.acquiring:
+            raise RuntimeError('Already counting')
+
+        self.acquiring = True
+        self.pvStart.put("Stop")
```
Comment (hhslepicka) [inline py4syn/epics/QE65000Class.py:154]: The device starts the acquisition when you write "Stop" to the Start PV?

File: py4syn/epics/QE65000Class.py [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+
+    def stopCount(self):
+        self.setCountStop()
+
+    def setPresetValue(self, channel, val):
+        """Dummy method"""
+        pass
+
+    # TODO: verificar
+    def close(self):
```
Comment (hhslepicka) [inline py4syn/epics/QE65000Class.py]: Isn't this the same as ```stopCount``` ?

File: py4syn/epics/QE65000Class.py [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+        """Stops an ongoing acquisition, if any, and puts the EPICS IOC in
+        idle state."""
+        pass
+
+    def startCollectImage(self, rows=0, cols=0):
```
Comment (hhslepicka) [inline py4syn/epics/QE65000Class.py]: Also the same as for DXP, maybe you can create a base class and DXP, DXPFake and this device can inherit from it. This way we have the code in a more maintainable fashion and easier to fix.

**Review by hhslepicka — CHANGES_REQUESTED** [review]
Hi @gabrielfedel! 
Thank you for your collaboration, nice work! :)
I made a few comments and some changes are needed before we can merge it.
Please let me know what you think about it.

File: py4syn/epics/QE65000Class.py (hunk 6 above)
Comment (gabrielfedel) [reply py4syn/epics/QE65000Class.py:154]: Yes... We are using an ioc from another author... I could change the ioc to let this better.

**Comment by gabrielfedel** [comment]
Great! I will review this soon! Thanks!

File: py4syn/epics/QE65000Class.py (hunk 8 above)
Comment (gabrielfedel) [reply py4syn/epics/QE65000Class.py]: Are you saying a base class to work with this collect image way?

File: py4syn/epics/QE65000Class.py (hunk 6 above)
Comment (hhslepicka) [reply py4syn/epics/QE65000Class.py:154]: No need to change the IOC, I was just curious about it! :+1:

File: py4syn/epics/QE65000Class.py (hunk 8 above)
Comment (hhslepicka) [reply py4syn/epics/QE65000Class.py]: Yes! In order to keep this code that is mainly the same but in one place. A base class or move this to something that can be imported and used by all of them and future developments with the same idea.

Comment (gabrielfedel) [reply py4syn/epics/QE65000Class.py]: great! I will do it.

File: py4syn/epics/QE65000Class.py [hunk 9]
```diff
@@ -42,7 +42,7 @@ def __init__(self, mnemonic, pv=None, responseTimeout=15, output="./out",
         self.pvDarkCorrection = PV(pv+":ElectricalDark")
 
         # the spectra come from different pv if use darkcorrection
-        if self.pvDarkCorrection.get() == "ON":
+        if self.pvDarkCorrection.get() == 1:
```
Comment (hhslepicka) [inline py4syn/epics/QE65000Class.py]: @gabrielfedel, is it possible to change during runtime or it is configured at the IOC and available as read-only on this PV?
If it is possible to change during runtime I would recommend to have some sort check on the moment that you read the pvSpectrum to point to the correct value.

File: py4syn/epics/DxpClass.py (hunk 1 above)
Comment (gabrielfedel) [reply py4syn/epics/DxpClass.py]: I change to numPoints

File: py4syn/epics/DxpClass.py (hunk 2 above)
Comment (gabrielfedel) [reply py4syn/epics/DxpClass.py]: solved

File: py4syn/epics/DxpClass.py (hunk 5 above)
Comment (gabrielfedel) [reply py4syn/epics/DxpClass.py]: Solved. I moved to ImageHDFClass

File: py4syn/epics/QE65000Class.py (hunk 9 above)
Comment (gabrielfedel) [reply py4syn/epics/QE65000Class.py]: Good idea

File: py4syn/epics/QE65000Class.py (hunk 7 above)
Comment (gabrielfedel) [reply py4syn/epics/QE65000Class.py]: yes, I removed it.

File: py4syn/epics/DxpClass.py (hunk 3 above)
Comment (gabrielfedel) [reply py4syn/epics/DxpClass.py]: The PVs has the format mcaC.Rr (where C is channel and R is ROI). If PV is only mcaC then we call saveSpectrum, if is a mcaC.Rr, is just a point collect.

I change to a constant this values (4 and 6)

File: py4syn/epics/DxpClass.py (hunk 4 above)
Comment (gabrielfedel) [reply py4syn/epics/DxpClass.py]: It is called on getValueChannel (when PV is a spectrum).

**Comment by gabrielfedel** [comment]
I think I finished the changes, tell me if you think necessary more updates.

File: py4syn/epics/ImageHDFClass.py [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+.. note:: 06/02/2017 [gabrielfedel]  first version released
+"""
+import os
+
+import numpy as np
+import h5py
+from py4syn.epics.StandardDevice import StandardDevice
+from py4syn.epics.ICountable import ICountable
+
+class ImageHDF(StandardDevice, ICountable):
```
Comment (hhslepicka) [inline py4syn/epics/ImageHDFClass.py:18]: ImageHDF should be just a class. Your devices must be StandardDevice and ICountable.
E.g.: DxpFake(StandardDevice, ICountable, ImageHDF).
Do you agree?

**Review by hhslepicka — CHANGES_REQUESTED** [review]

**Comment by hhslepicka** [comment]
Hi @gabrielfedel, thank you for the modifications. It is getting much better!
There is just this one point that I mentioned on the review. Let me know if you agree.

**Comment by hhslepicka** [comment]
Also, please squash it in order to facilitate the review and history.

File: py4syn/epics/ImageHDFClass.py (hunk 10 above)
Comment (gabrielfedel) [reply py4syn/epics/ImageHDFClass.py:18]: Hello Hugo. My first try was like your suggestion, but i not found a way to start the 2 constructors (StandardDevice and ImageHDF) on subclasses... Do you know if there is a way to do this?  Or, maybe, another solution (remove from ImageHDF constructor and create a simple method)...

**Comment by gabrielfedel** [comment]
Sorry, I didn't understand it "lease squash it in order to facilitate the review and history". Did I do something in a wrong order?

File: py4syn/epics/ImageHDFClass.py (hunk 10 above)
Comment (hhslepicka) [reply py4syn/epics/ImageHDFClass.py:18]: Hi Gabriel, I will take a better look at that! For now let's leave as is.

**Comment by hhslepicka** [comment]
@gabrielfedel, squash is used to reduce the number of commits and make it easier to review and also check the history.
Please take a look at: http://stackoverflow.com/questions/31668794/squash-all-your-commits-in-one-before-a-pull-request-in-github

**Comment by gabrielfedel** [comment]
Thank you for the reference. I will use squash next time. 
Something else to fix?

**Comment by hhslepicka** [comment]
@gabrielfedel, looks good to me now.
Thank you!

**Review by hhslepicka — APPROVED** [review]

**Comment by gabrielfedel** [comment]
Great!
Thank you for review!

**Pull request merged**